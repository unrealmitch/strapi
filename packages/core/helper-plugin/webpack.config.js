const webpack = require('webpack');
const { ESBuildMinifyPlugin } = require('esbuild-loader');
const browserslistToEsbuild = require('browserslist-to-esbuild');

const packageJson = require('./package.json');

const nodeModules = [];
[...Object.keys(packageJson.dependencies), ...Object.keys(packageJson.peerDependencies)].forEach(
  (module) => {
    nodeModules.push(new RegExp(`^${module}(/.+)?$`));
  }
);

/** @type {Omit<import('webpack').Configuration, 'output'>} */
const baseConfig = {
  entry: `${__dirname}/lib/src/index.js`,
  externals: nodeModules,
  mode: process.env.NODE_ENV,
  devtool: process.env.NODE_ENV === 'production' ? false : 'eval-source-map',
  optimization: {
    minimize: process.env.NODE_ENV === 'production',
    minimizer: [
      new ESBuildMinifyPlugin({
        target: browserslistToEsbuild(),
      }),
    ],
  },
  module: {
    rules: [
      {
        test: /\.m?jsx?$/,
        use: {
          loader: require.resolve('esbuild-loader'),
          options: {
            loader: 'jsx',
            target: browserslistToEsbuild(),
          },
        },
      },
      {
        test: /\.(png|svg|jpg|gif)$/,
        type: 'asset',
        parser: {
          dataUrlCondition: {
            maxSize: 8192,
          },
        },
      },
    ],
  },
  resolve: {
    extensions: ['*', '.js'],
    cacheWithContext: false,
  },
  plugins: [
    new webpack.EnvironmentPlugin({
      NODE_ENV: 'production',
    }),
  ],
};

/** @type {import('webpack').Configuration[]} */
const config = [
  {
    ...baseConfig,
    output: {
      path: `${__dirname}/build`,
      filename: `helper-plugin.esm.js`,
      library: {
        type: 'module',
      },
      environment: { module: true },
    },
    experiments: {
      outputModule: true,
    },
  },
  {
    ...baseConfig,
    output: {
      path: `${__dirname}/build`,
      filename: `helper-plugin.js`,
      library: {
        type: 'commonjs',
      },
    },
  },
];

module.exports = config;
