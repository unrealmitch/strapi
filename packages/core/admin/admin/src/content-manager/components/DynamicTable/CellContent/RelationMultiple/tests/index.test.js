import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { ThemeProvider, lightTheme } from '@strapi/design-system';
import { IntlProvider } from 'react-intl';
import { QueryClientProvider, QueryClient } from 'react-query';

import { useFetchClient } from '@strapi/helper-plugin';
import RelationMultiple from '../index';

jest.mock('@strapi/helper-plugin', () => ({
  ...jest.requireActual('@strapi/helper-plugin'),
  useFetchClient: jest.fn().mockReturnValue({
    get: jest.fn().mockResolvedValue({
      data: {
        results: [
          {
            id: 1,
            name: 'Relation entity 1',
          },
          {
            id: 2,
            name: 'Relation entity 2',
          },
          {
            id: 3,
            name: 'Relation entity 3',
          },
        ],

        pagination: {
          total: 1,
        },
      },
    }),
  }),
}));

const DEFAULT_PROPS_FIXTURE = {
  contentType: {
    uid: 'api::address.address',
  },
  entityId: 1,
  fieldSchema: {
    type: 'relation',
    relation: 'manyToMany',
    target: 'api::category.category',
  },
  metadatas: {
    mainField: {
      name: 'name',
      schema: {
        type: 'string',
      },
    },
  },
  value: {
    count: 1,
  },
  name: 'categories.name',
  rowId: 1,
};

const ComponentFixture = () => {
  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en" messages={{}} defaultLocale="en">
          <RelationMultiple {...DEFAULT_PROPS_FIXTURE} />
        </IntlProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

describe('DynamicTable / Cellcontent / RelationMultiple', () => {
  it('renders and matches the snapshot', async () => {
    const { container } = render(<ComponentFixture />);
    expect(container).toMatchSnapshot();
    const { get } = useFetchClient();
    expect(get).toHaveBeenCalledTimes(0);
  });

  it('fetches relation entities once the menu is opened', async () => {
    const { container } = render(<ComponentFixture />);
    const { get } = useFetchClient();
    const button = container.querySelector('[type=button]');

    fireEvent(button, new MouseEvent('mousedown', { bubbles: true }));

    expect(screen.getByText('Relations are loading')).toBeInTheDocument();
    expect(get).toHaveBeenCalledTimes(1);
    await waitFor(() => expect(screen.getByText('Relation entity 1')).toBeInTheDocument());
  });

  it('Displays related entities in reversed order', async () => {
    const { container } = render(<ComponentFixture />);
    const button = container.querySelector('[type=button]');

    fireEvent(button, new MouseEvent('mousedown', { bubbles: true }));

    await waitFor(() => {
      const buttons = screen.getAllByRole('menuitem');

      expect(buttons[1]).toHaveTextContent('Relation entity 3');
      expect(buttons[2]).toHaveTextContent('Relation entity 2');
      expect(buttons[3]).toHaveTextContent('Relation entity 1');
    });
  });
});
