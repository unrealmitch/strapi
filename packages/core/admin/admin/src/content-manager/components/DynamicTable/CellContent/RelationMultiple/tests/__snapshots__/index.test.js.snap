// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicTable / Cellcontent / RelationMultiple renders and matches the snapshot 1`] = `
.c0 {
  background: #4945ff;
  padding: 8px;
  padding-right: 16px;
  padding-left: 16px;
  border-radius: 4px;
  border-color: #4945ff;
  border: 1px solid #4945ff;
  cursor: pointer;
}

.c8 {
  background: #eaeaef;
  padding-right: 8px;
  padding-left: 8px;
  min-width: 20px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 8px;
}

.c7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  gap: 4px;
}

.c9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #ffffff;
}

.c11 {
  font-weight: 600;
  font-size: 0.6875rem;
  line-height: 1.45;
  text-transform: uppercase;
  color: #666687;
}

.c10 {
  border-radius: 4px;
  height: 1.5rem;
}

.c2 {
  position: relative;
  outline: none;
}

.c2 svg {
  height: 12px;
  width: 12px;
}

.c2 svg > g,
.c2 svg path {
  fill: #ffffff;
}

.c2[aria-disabled='true'] {
  pointer-events: none;
}

.c2:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c2:focus-visible {
  outline: none;
}

.c2:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c14 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  height: 2rem;
  border: 1px solid transparent;
  background: transparent;
}

.c3[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c3[aria-disabled='true'] .c5 {
  color: #666687;
}

.c3[aria-disabled='true'] svg > g,.c3[aria-disabled='true'] svg path {
  fill: #666687;
}

.c3[aria-disabled='true']:active {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c3[aria-disabled='true']:active .c5 {
  color: #666687;
}

.c3[aria-disabled='true']:active svg > g,.c3[aria-disabled='true']:active svg path {
  fill: #666687;
}

.c3:hover {
  background-color: #f6f6f9;
}

.c3:active {
  border: 1px solid undefined;
  background: undefined;
}

.c3 .c5 {
  color: #32324d;
}

.c3 svg > g,
.c3 svg path {
  fill: #8e8ea9;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 svg {
  height: 4px;
  width: 6px;
}

.c4 {
  padding: 4px 12px;
}

<div>
  <div
    aria-hidden="true"
    class=""
    role="button"
  >
    <div>
      <button
        aria-controls="0"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="true"
        class="c0 c1 c2 c3 c4"
        type="button"
      >
        <span
          class="c5 c6"
        >
          <div
            class="c7"
            wrap="nowrap"
          >
            <div
              class="c8 c9 c10"
            >
              <span
                class="c5 c11"
              >
                1
              </span>
            </div>
            item
          </div>
        </span>
        <div
          aria-hidden="true"
          class="c12"
        >
          <span
            class="c13"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="1rem"
              viewBox="0 0 14 8"
              width="1rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14 .889a.86.86 0 0 1-.26.625L7.615 7.736A.834.834 0 0 1 7 8a.834.834 0 0 1-.615-.264L.26 1.514A.861.861 0 0 1 0 .889c0-.24.087-.45.26-.625A.834.834 0 0 1 .875 0h12.25c.237 0 .442.088.615.264a.86.86 0 0 1 .26.625Z"
                fill="#32324D"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </button>
    </div>
  </div>
  <div
    class="c14"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
